### PR TITLE
Add recipe output for r-tiledbsoma

### PR DESCRIPTION
**Pull request by jdblischak**
This PR adds an output section to build the conda package r-tiledbsoma

* Good news: I was able to build the recipe locally on Ubuntu for R 4.1 and 4.2. I've copied some relevant output from the successful build below

* Bad news: The Azure builds fail. For some strange reason, the conda solver tries to install R 3.5. Still have to investigate

* Temporary changes: note that I made a few temporary changes to facilitate testing. First I removed all but one of the Python variants to reduce the build time. Second I switched to use the Git repo as the source since enabling the conda builds is an ongoing process

* Immediate question: The Python client version exactly matches the libtiledbsoma version. Is the plan to keep a separate version string for the R client?

cc: @eddelbuettel, @nguyenv, @aaronwolen, @ihnorton 
xref: https://github.com/single-cell-data/TileDB-SOMA/issues/1277#issuecomment-1520930305, https://github.com/single-cell-data/TileDB-SOMA/pull/1302

Excerpts from my successful local build:

```sh
> tiledbsoma::show_package_versions()
tiledbsoma:    0.0.0.9022
tiledb-r:      0.19.1
tiledb core:   2.15.2
libtiledbsoma: libtiledb=2.15.2
R:             R version 4.1.3 (2022-03-10)
OS:            Ubuntu 22.04.2 LTS

> tiledbsoma::show_package_versions()
tiledbsoma:    0.0.0.9022
tiledb-r:      0.19.1
tiledb core:   2.15.2
libtiledbsoma: libtiledb=2.15.2
R:             R version 4.2.3 (2023-03-15)
OS:            Ubuntu 22.04.2 LTS
```

```json
{
  "libtiledbsoma-1.2.2-h238ffd6_0": {
    "recipe": {
      "c_compiler": "gcc",
      "c_compiler_version": "12",
      "channel_targets": "conda-forge main",
      "cxx_compiler": "gxx",
      "cxx_compiler_version": "12",
      "fmt": "9",
      "spdlog": "1.11",
      "target_platform": "linux-64"
    }
  },
  "r-tiledbsoma-0.0.0.9022-r41h00ab1b0_0": {
    "recipe": {
      "channel_targets": "conda-forge main",
      "cxx_compiler": "gxx",
      "cxx_compiler_version": "12",
      "target_platform": "linux-64"
    }
  },
  "r-tiledbsoma-0.0.0.9022-r42h00ab1b0_0": {
    "recipe": {
      "channel_targets": "conda-forge main",
      "cxx_compiler": "gxx",
      "cxx_compiler_version": "12",
      "target_platform": "linux-64"
    }
  },
  "tiledbsoma-py-1.2.2-py310hfb6f7a9_0": {
    "recipe": {
      "channel_targets": "conda-forge main",
      "cxx_compiler": "gxx",
      "cxx_compiler_version": "12",
      "numpy": "1.21",
      "python": "3.10.* *_cpython",
      "target_platform": "linux-64"
    }
  }
}
```

### CODE REVIEW

**Comment by jdblischak** [comment]
Duh, I forgot to rerender. Thus it didn't use the conda-forge-pinning of r-base 4.1 and 4.2. Locally I was manually pointing to a copy of conda-forge-pinning, so that's why it had worked for me

**Comment by jdblischak** [comment]
Current status:

* linux-64 build passed
* osx-64 build [failed](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=31944&view=logs&j=58ac6aab-c4bc-5de2-2894-98e408cc8ec9&t=933f325c-924e-533d-4d95-e93b5843ce8b&l=956) during compilation
  ```sh
  In file included from ../inst/tiledb/include/tiledb/attribute.h:42:
  ../inst/tiledb/include/tiledb/object.h:131:30: error: 'value' is unavailable: introduced in macOS 10.13 - see https://conda-forge.org/docs/maintainer/knowledge_base.html#newer-c-features-with-old-sdk
      ret += " - \"" + name_.value() + "\">";
  ```
* osx-arm64 build [failed](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=31944&view=logs&j=ced5d8de-8227-5f3f-33a1-45cf1592c45a&t=f7bcb349-6dca-53ba-d88d-1d9963e119ba&l=1594) during byte-compliation

  ```
  ** byte-compile and prepare package for lazy loading
  Error in dyn.load(file, DLLpath = DLLpath, ...) : 
    unable to load shared object '/Users/runner/miniforge3/conda-bld/tiledbsoma_1683311692471/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/lattice/libs/lattice.dylib':
    dlopen(/Users/runner/miniforge3/conda-bld/tiledbsoma_1683311692471/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/lattice/libs/lattice.dylib, 6): no suitable image found.  Did find:
  	/Users/runner/miniforge3/conda-bld/tiledbsoma_1683311692471/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/lattice/libs/lattice.dylib: mach-o, but wron
  Calls: <Anonymous> ... asNamespace -> loadNamespace -> library.dynam -> dyn.load
  Execution halted
  ERROR: lazy loading failed for package ‘tiledbsoma’
  ```

**Comment by johnkerl** [comment]
@jdblischak the `unable to load shared object` might be addressed by https://github.com/single-cell-data/TileDB-SOMA/pull/1343 which is a WIP

**Comment by Shelnutt2** [comment]
> @jdblischak the `unable to load shared object` might be addressed by [single-cell-data/TileDB-SOMA#1343](https://github.com/single-cell-data/TileDB-SOMA/pull/1343) which is a WIP

I don't think that PR will help, the error is not for `libtiledbsoma` but for `lattice.dylib`. Also conda doesn't use `/usr/local` as a path, libraries are in the conda prefix location, which pkg-config should have set from the c++ library.

@jdblischak nice work, thanks for putting this up! I'm wondering if the `r-lattice` is broken upstream?

For the amd64 issue I think we'll need to wrap the compiler commands like we did in `r-tiledb`: https://github.com/conda-forge/r-tiledb-feedstock/blob/658080fdbc2ed3c3a0a5039ca4e30460f005c26b/recipe/build.sh#L4

**Comment by eddelbuettel** [comment]
> I'm wondering if the r-lattice is broken upstream?

Unlikely, it is a 'recommended' package that 'effectively' comes with R itself (though it is technically a CRAN package like 19.5k others).  But maybe the Conda package needs a rebuild around the R 4.2.3 to R 4.3.0 transition?

**Comment by jdblischak** [comment]
Thanks all for the feedback. The osx-64 build is passing now!

> I'm wondering if the r-lattice is broken upstream?

@Shelnutt2 This was a promising lead. There were previous osx-arm64 conda binaries that were actually osx-64 (https://github.com/conda-forge/r-lattice-feedstock/issues/14). However, these have been marked as broken (https://github.com/conda-forge/admin-requests/pull/218), and the recipe subsequently fixed (https://github.com/conda-forge/r-lattice-feedstock/pull/15). I installed the same osx-arm64 r-lattice versions from the Azure build on a local macOS M1 machine, and I was able to run example lattice code without issue

**Comment by jdblischak** [comment]
Latest update. I've made progress on the osx-arm64 build. By including all the R package dependencies that require compilation in the cross-compilation build requirements, I was able to remove the loading errors from unrelated packages (eg `lattice.dylib`).

However, now it is [failing](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=31986&view=logs&j=ced5d8de-8227-5f3f-33a1-45cf1592c45a&t=f7bcb349-6dca-53ba-d88d-1d9963e119ba&l=1629) to load `tiledbsoma.dylib` itself, which was surprising given that it should have just been built. 

```
** byte-compile and prepare package for lazy loading
** help
Error :  unable to load shared object '/Users/runner/miniforge3/conda-bld/tiledbsoma_1683650279834/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/00LOCK-r/00new/tiledbsoma/libs/tiledbsoma.dylib':
  dlopen(/Users/runner/miniforge3/conda-bld/tiledbsoma_1683650279834/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/00LOCK-r/00new/tiledbsoma/libs/tiledbsoma.dylib, 6): no suitable image found.  Did find:
	/Users/runner/miniforge3/conda-bld/tiledbsoma_1683650279834/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/00LOCK-r/0
ERROR: installing Rd objects failed for package ‘tiledbsoma’
* removing ‘/Users/runner/miniforge3/conda-bld/tiledbsoma_1683650279834/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/tiledbsoma’
```

Looking at the recipes for [r-tiledb](https://github.com/conda-forge/r-tiledb-feedstock/blob/658080fdbc2ed3c3a0a5039ca4e30460f005c26b/recipe/meta.yaml) and [tiledbsoma-py](https://github.com/TileDB-Inc/tiledbsoma-feedstock/blob/0d5197f02f006c7c13bcc4c3440fbbdf5b129e93/recipe/meta.yaml#L49), I don't see anything overly obvious that I am doing wrong. Any suggestions?

**Comment by eddelbuettel** [comment]
The only difference I know of between macOS arm64 and macOS x86_64 is the imposition of the minimum-compilation standard for x86_64 _only_ to permit use on previous releases.  

Could that matter here, i.e. could we accidentally have -mmacosx-version-min=10.14 on for arm64 ?  Otherwise I am more than stumped.

**Comment by jdblischak** [comment]
> Could that matter here, i.e. could we accidentally have -mmacosx-version-min=10.14 on for arm64 ? Otherwise I am more than stumped.

Thanks Dirk. If I understand your suggestion correctly, I think the answer is no. The supplemental scripts that set `-mmacosx-version-min=10.14` are only applied when the conda-build env var `target_platform` is equal to "osx-64". In the arm64 build it is set to "osx-arm64", and thus that code is skipped.

Checking the [build log](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=31986&view=logs&j=ced5d8de-8227-5f3f-33a1-45cf1592c45a&t=f7bcb349-6dca-53ba-d88d-1d9963e119ba&l=1313) to confirm:

```
+ cd apis/r
+ export DISABLE_AUTOBREW=1
+ DISABLE_AUTOBREW=1
+ [[ osx-arm64 == osx-64 ]]
+ export 'CXX17FLAGS=-Wno-deprecated-declarations -Wno-deprecated'
+ CXX17FLAGS='-Wno-deprecated-declarations -Wno-deprecated'
+ /Users/runner/miniforge3/conda-bld/tiledbsoma_1683650279834/_build_env/bin/R CMD INSTALL --build . --library=/Users/runner/miniforge3/conda-bld/tiledbsoma_1683650279834/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library --no-test-load
* installing *source* package ‘tiledbsoma’ ...
** using staged installation
```

**Comment by jdblischak** [comment]
Since there have been multiple [updates](https://github.com/single-cell-data/TileDB-SOMA/commits/main/apis/r) to the R client since last week, I restarted the previously failed osx-arm64 build. This [new build](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=31986&view=logs&j=d6e17299-69de-5342-9bfa-d2a6718dd363&t=07f994f3-c7f4-5985-4d61-e77794d5f611) used the TileDB-SOMA source repo from commit [18f0394](https://github.com/single-cell-data/TileDB-SOMA/tree/18f0394d1e36295db94bd6724354af02abd674a3)

The [error](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=31986&view=logs&j=d6e17299-69de-5342-9bfa-d2a6718dd363&t=07f994f3-c7f4-5985-4d61-e77794d5f611&l=1526) this time is similar, though it includes the name of the help file that was being processed when the error was triggered, [`EphemeralCollection.Rd`](https://github.com/single-cell-data/TileDB-SOMA/blob/main/apis/r/man/EphemeralCollection.Rd). Assuming that the help files are processed alphabetically, this means that [`ConfigList.Rd`](https://github.com/single-cell-data/TileDB-SOMA/blob/main/apis/r/man/ConfigList.Rd) was able to be processed without error

```
** byte-compile and prepare package for lazy loading
** help
Error : EphemeralCollection.Rd:7: unable to load shared object '/Users/runner/miniforge3/conda-bld/tiledbsoma_1684257316766/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/00LOCK-r/00new/tiledbsoma/libs/tiledbsoma.dylib':
  dlopen(/Users/runner/miniforge3/conda-bld/tiledbsoma_1684257316766/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/00LOCK-r/00new/tiledbsoma/libs/tiledbsoma.dylib, 6): no suitable image found.  Did find:
	/Users/runner/miniforge3/conda-bld/tiledbsoma_1684257316766/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/00LOCK-r/0
ERROR: installing Rd objects failed for package ‘tiledbsoma’
* removing ‘/Users/runner/miniforge3/conda-bld/tiledbsoma_1684257316766/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/tiledbsoma’
```

File: recipe/build-r-tiledbsoma.sh [hunk 1]
```diff
@@ -20,4 +20,4 @@ fi
 
 export CXX17FLAGS="-Wno-deprecated-declarations -Wno-deprecated"
 
-${R} CMD INSTALL --build . ${R_ARGS}
+${R} CMD INSTALL --build --no-help . ${R_ARGS}
```
Comment (johnkerl) [inline recipe/build-r-tiledbsoma.sh:23]: @eddelbuettel @aaronwolen @mojaveazure may have thoughts here ... from my own (relatively R-amateur) understanding I tend to use:

```
alias rcb='rm -f *.tar.gz; R CMD build --no-build-vignettes --no-manual .'
alias rci='R CMD INSTALL --no-test-load --no-docs --no-html *tar.gz'
```

for interactive use ...

Comment (eddelbuettel) [reply recipe/build-r-tiledbsoma.sh:23]: I fear that if that is in regards to the build issue the help files are likely just the messenger you both may be aiming and firing at the messenger.

The help files have, as we are so fancy, a couple of call-outs to an `\Sexpr{}` leading to the loading of the package to evaluate the expression.  So we "just" fail to load the package earlier than we would at the next step.

I think this is symptomatic of an rpath issue.  We have one on non-Conda licked.  Are there directories that are different from non-Conda builds that we need to set an rpath for?

Comment (jdblischak) [reply recipe/build-r-tiledbsoma.sh:23]: To clarify, adding `--no-help` was a temporary workaround to trigger a different error message later in the package installation process. To my surprise, the [installation succeeded on osx-arm64](https://dev.azure.com/jdblischak/feedstock-builds/_build/results?buildId=278&view=logs&j=ced5d8de-8227-5f3f-33a1-45cf1592c45a&t=f7bcb349-6dca-53ba-d88d-1d9963e119ba&l=3719) when the help file installation step was skipped. However, the test section is skipped for cross-compiled builds, so I'm still somewhat skeptical that the package is functional, but it's a positive sign

```
+ cd apis/r
+ export DISABLE_AUTOBREW=1
+ DISABLE_AUTOBREW=1
+ [[ osx-arm64 == osx-64 ]]
+ export 'CXX17FLAGS=-Wno-deprecated-declarations -Wno-deprecated'
+ CXX17FLAGS='-Wno-deprecated-declarations -Wno-deprecated'
+ /Users/runner/miniforge3/conda-bld/tiledbsoma_1684260630295/_build_env/bin/R CMD INSTALL --build --no-help . --library=/Users/runner/miniforge3/conda-bld/tiledbsoma_1684260630295/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library --no-test-load
* installing *source* package ‘tiledbsoma’ ...
** using staged installation
Error: No 'cmake' binary found
Execution halted
./configure: line 44: tools/build_libtiledbsoma.sh: No such file or directory
** libs
if [ `uname -s` = 'Darwin' ] && [ -f tiledbsoma.so ]; then \
    if [ -f ../inst/tiledb/lib/libtiledb.dylib ] ; then \
        install_name_tool -change libz.1.dylib @rpath/libz.1.dylib ../inst/tiledb/lib/libtiledb.dylib; \
        install_name_tool -add_rpath @loader_path/../tiledb/lib tiledbsoma.so; \
        install_name_tool -add_rpath @loader_path/../tiledbsoma/lib tiledbsoma.so; \
    fi; \
    if [ -f /usr/local/lib/libtiledb.dylib -a -f /usr/local/lib/libtiledbsoma.dylib ] ; then \
        install_name_tool -add_rpath /usr/local/lib tiledbsoma.so; \
** checking absolute paths in shared objects and dynamic libraries
* creating tarball
packaged installation of ‘tiledbsoma’ as ‘tiledbsoma_0.0.0.9024.tgz’
* DONE (tiledbsoma)
```

Comment (jdblischak) [reply recipe/build-r-tiledbsoma.sh:23]: I was able to remove the `cmake` not found error by installing `pkg-config` as a build dependency. However, that broke the osx-64 (ie amd64) build. I've seen this error about `Library not loaded: @rpath/R/lib/libR.dylib` before. It's either spurious or related to some of the upstream changes you've been making. I have some ideas to try

```
** R
** inst
** byte-compile and prepare package for lazy loading
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
Error: package or namespace load failed for ‘tiledbsoma’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/Users/runner/miniforge3/conda-bld/tiledbsoma_1684266988093/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/00LOCK-r/00new/tiledbsoma/libs/tiledbsoma.dylib':
  dlopen(/Users/runner/miniforge3/conda-bld/tiledbsoma_1684266988093/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/00LOCK-r/00new/tiledbsoma/libs/tiledbsoma.dylib, 6): Library not loaded: @rpath/R/lib/libR.dylib
  Referenced from: /Users/runner/miniforge3/conda-bld/tiledbsoma_1684266988093/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/Users/runner/miniforge3/conda-bld/tiledbsoma_1684266988093/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehol/lib/R/library/tiledbsoma’
```

Comment (eddelbuettel) [reply recipe/build-r-tiledbsoma.sh]: (Scratching my head:  libR.dylib is what we get, that's from what would be `r-base` in Ubuntu. Conda not loading its own package? Hm.)

Comment (jdblischak) [reply recipe/build-r-tiledbsoma.sh]: My current hunch is that the calls to `install_name_tool` are messing with the standard conda `rpath` settings, which is why things break when trying to load it from a different location. But I have a plan to explicitly test this

**Comment by jdblischak** [comment]
My troubleshooting progress was slow today. My ideas related to editing `configure` and `Makevars.in` for the conda build were a dead end, and seemed to hurt more than help anything.

One useful thing I did was decrease the installation prefix length so that we can see the full error messages from the macOS builds (see the conda build docs on [making packages relocatable](https://docs.conda.io/projects/conda-build/en/stable/resources/make-relocatable.html) for the justification for using the long installation prefix).

For the osx-64 build (macOS amd64), the build works with R 4.2. The failure comes when trying to build with R 4.1. For some reason, `tiledbsoma.dylib` appears to require R 4.2+

```
installing to /Users/runner/miniforge3/conda-bld/tiledbsoma_1684350979543/_h_env_placehold_placehold_placehold_placehold_placehold_pla/lib/R/library/00LOCK-r/00new/tiledbsoma/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
Error :  unable to load shared object '/Users/runner/miniforge3/conda-bld/tiledbsoma_1684350979543/_h_env_placehold_placehold_placehold_placehold_placehold_pla/lib/R/library/00LOCK-r/00new/tiledbsoma/libs/tiledbsoma.dylib':
  dlopen(/Users/runner/miniforge3/conda-bld/tiledbsoma_1684350979543/_h_env_placehold_placehold_placehold_placehold_placehold_pla/lib/R/library/00LOCK-r/00new/tiledbsoma/libs/tiledbsoma.dylib, 6): Library not loaded: @rpath/R/lib/libR.dylib
  Referenced from: /Users/runner/miniforge3/conda-bld/tiledbsoma_1684350979543/_h_env_placehold_placehold_placehold_placehold_placehold_pla/lib/R/library/00LOCK-r/00new/tiledbsoma/libs/tiledbsoma.dylib
  Reason: Incompatible library version: tiledbsoma.dylib requires version 4.2.0 or later, but libR.dylib provides version 4.1.0
ERROR: installing Rd objects failed for package ‘tiledbsoma’
* removing ‘/Users/runner/miniforge3/conda-bld/tiledbsoma_1684350979543/_h_env_placehold_placehold_placehold_placehold_placehold_pla/lib/R/library/tiledbsoma’
```

The key error message is:

> Reason: Incompatible library version: tiledbsoma.dylib requires version 4.2.0 or later, but libR.dylib provides version 4.1.0

I couldn't figure out where this requirement comes from. The `DESCRIPTION` file states `R (>= 4.0.0)`. Any insights?

For the osx-arm64 build, it is able to find `libs/tiledbsoma.dylib`, but complains that it is the wrong architecture (`mach-o, but wrong architecture`). To me this strongly implies there is something wrong with the cross-compilation.

```
installing to /Users/runner/miniforge3/conda-bld/tiledbsoma_1684350977931/_h_env_placehold_placehold_placehold_placehold_placehold_pla/lib/R/library/00LOCK-r/00new/tiledbsoma/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
Error :  unable to load shared object '/Users/runner/miniforge3/conda-bld/tiledbsoma_1684350977931/_h_env_placehold_placehold_placehold_placehold_placehold_pla/lib/R/library/00LOCK-r/00new/tiledbsoma/libs/tiledbsoma.dylib':
  dlopen(/Users/runner/miniforge3/conda-bld/tiledbsoma_1684350977931/_h_env_placehold_placehold_placehold_placehold_placehold_pla/lib/R/library/00LOCK-r/00new/tiledbsoma/libs/tiledbsoma.dylib, 6): no suitable image found.  Did find:
	/Users/runner/miniforge3/conda-bld/tiledbsoma_1684350977931/_h_env_placehold_placehold_placehold_placehold_placehold_pla/lib/R/library/00LOCK-r/00new/tiledbsoma/libs/tiledbsoma.dylib: mach-o, but wrong architecture
	/Users/runner/miniforge3/conda-bld/tiledbsoma_1684350977931/_h_env_placehold_placehold_placehold_placehold_placehold_pla/lib/R/library/00LOCK-r/00new/tiledbsoma/libs/tiledbsoma.dylib: mach-o, but wrong architecture
ERROR: installing Rd objects failed for package ‘tiledbsoma’
* removing ‘/Users/runner/miniforge3/conda-bld/tiledbsoma_1684350977931/_h_env_placehold_placehold_placehold_placehold_placehold_pla/lib/R/library/tiledbsoma’
```

In summary, I don't think this is an issue with `rpath` or in general finding the libraries. The problem is that the `tiledbsoma.dylib` that is created and found is incompatible for some reason.

**Comment by jdblischak** [comment]
This osx-64 (macOS x86) build error is so frustrating. Once again the error has disappeared despite the fact that I didn't change anything relevant. Now it [builds](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=32112&view=logs&j=58ac6aab-c4bc-5de2-2894-98e408cc8ec9&t=933f325c-924e-533d-4d95-e93b5843ce8b) just fine with R 4.1.3, but next time is anyone's guess

```
INFO (r-tiledbsoma,lib/R/library/tiledbsoma/libs/tiledbsoma.dylib): Needed DSO lib/R/lib/libR.dylib found in conda-forge::r-base-4.1.3-h7a6543b_8
```

Actually, I have a hunch. I think the build order (R 4.1 vs 4.2) is playing a role. The order is randomly decided for each job. This most recent one [passed](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=32112&view=logs&j=58ac6aab-c4bc-5de2-2894-98e408cc8ec9&t=933f325c-924e-533d-4d95-e93b5843ce8b&l=69) when the R 4.1 build preceded the R 4.2 build:

```
BUILD START: ['libtiledbsoma-1.2.2-h2650b1e_0.conda', 'r-tiledbsoma-0.0.0.9024-r41hfb802ea_0.conda', 'r-tiledbsoma-0.0.0.9024-r42hfb802ea_0.conda', 'tiledbsoma-py-1.2.2-py39hc51d931_0.conda']
```

And yesterday's build [failed](https://dev.azure.com/TileDB-Inc/CI/_build/results?buildId=32094&view=logs&j=58ac6aab-c4bc-5de2-2894-98e408cc8ec9&t=933f325c-924e-533d-4d95-e93b5843ce8b&l=67) when the R 4.2 build was first:

```
BUILD START: ['libtiledbsoma-1.2.2-h2650b1e_0.conda', 'r-tiledbsoma-0.0.0.9022-r42hfb802ea_0.conda', 'r-tiledbsoma-0.0.0.9022-r41hfb802ea_0.conda', 'tiledbsoma-py-1.2.2-py39hc51d931_0.conda']
```

This suggests that something from the 4.2 build is leaking outside of the conda env, and is then found again during the 4.1 build

**Comment by jdblischak** [comment]
Alright. I finally figured out the problem with the osx-64 build, that technically also applied to the linux-64 build even though it didn't cause an error. The compiled files are saved in the source directory, and they weren't being cleaned up between builds, and thus they weren't being recompiled with the next version of R. The fix was adding the flag `--clean`

File: recipe/build-r-tiledbsoma.sh [hunk 2]
```diff
@@ -20,4 +20,16 @@ fi
 
 export CXX17FLAGS="-Wno-deprecated-declarations -Wno-deprecated"
 
-${R} CMD INSTALL --preclean --clean --build . ${R_ARGS}
+# Unlike most R recipes which are built for one R version per job, this recipe
```
Comment (johnkerl) [inline recipe/build-r-tiledbsoma.sh:23]: ❤️

**Comment by jdblischak** [comment]
Just as the osx-64 build was fixed with the deceptively simple `--clean`, the osx-arm64 build was fixed with `--no-help`. I had already observed this above, but I didn't understand why it worked. Now I understand why. The [help pages in TileDB-SOMA-R](https://github.com/search?q=repo%3Asingle-cell-data%2FTileDB-SOMA+sexpr&type=code) use the macro `\Sexpr{}` to [dynamically evaluate R code](https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Dynamic-pages) when rendering the help pages. I didn't have to use `--no-help` when cross-compiling TileDB-R because it doesn't use `\Sexpr{}` in its [help pages](https://github.com/search?q=repo%3ATileDB-Inc%2FTileDB-R%20sexpr&type=code).

**Comment by jdblischak** [comment]
After the next TileDB-SOMA release, I'll rebase, cleanup, and remove the draft status

**Comment by johnkerl** [comment]
Thanks @jdblischak !!! :)

**Comment by jdblischak** [comment]
Superseded by #24. The R client output was included in the 1.2.4 release
